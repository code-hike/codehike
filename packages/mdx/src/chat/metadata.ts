export function setMetadata(
  content: string,
  metadata: Record<string, string>
) {
  const oldData = getMetadata(content)

  const metadataString = Object.entries({
    ...oldData,
    ...metadata,
  })
    .map(([key, value]) => `${key}: ${value}`)
    .join("\n")

  const contentWithoutMetadata = removeMetadata(content)

  return `---\n${metadataString}\n---\n${contentWithoutMetadata}`
}

export function getMetadata(
  content: string
): Record<string, string> {
  if (!content) {
    return undefined
  }
  const metadata = content.match(/^---\n([\s\S]*?)\n---\n/)

  if (!metadata) {
    return {}
  }

  const metadataString = metadata[1]

  const metadataEntries = metadataString
    .split("\n")
    .map(s => s.split(": "))
    .filter(([key, value]) => key && value)

  return Object.fromEntries(metadataEntries)
}

export function removeMetadata(content: string) {
  return content.replace(/^---\n[\s\S]*?\n---\n/, "")
}
