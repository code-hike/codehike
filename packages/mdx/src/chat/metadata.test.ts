import { expect, test } from "vitest"
import { setMetadata } from "./metadata"

test("set", () => {
  expect(
    setMetadata("hello", { foo: "bar" })
  ).toMatchSnapshot()
})

test("add", () => {
  const content = setMetadata("hello", { foo: "bar" })
  expect(setMetadata(content, { x: "y" })).toMatchSnapshot()
})

test("change", () => {
  const content = setMetadata("hello", { foo: "bar" })
  expect(
    setMetadata(content, { foo: "y" })
  ).toMatchSnapshot()
})
