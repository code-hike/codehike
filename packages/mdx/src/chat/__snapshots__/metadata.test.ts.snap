// Vitest Snapshot v1

exports[`add 1`] = `
"---
foo: bar
x: y
---
hello"
`;

exports[`change 1`] = `
"---
foo: y
---
hello"
`;

exports[`set 1`] = `
"---
foo: bar
---
hello"
`;
